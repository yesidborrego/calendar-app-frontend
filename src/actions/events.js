import Swal from "sweetalert2";

import { convertDateEvents } from "../helpers/convertDateEvents";
import { fetchWithToken } from "../helpers/fetch";
import { types } from "../types/types";

export const eventSetActive = (event) => (
  {
    type: types.eventSetActive,
    payload: event
  }
);

export const eventStartAddNew = (event) => {
  return async (dispatch, getState) => {
    const { uid, name } = getState().auth;

    try {
      const res = await fetchWithToken('events', event, 'POST');
      const body = await res.json();

      event.id = body.evento.id;
      event.user = {
        _id: uid,
        name
      };

      dispatch(eventAddNew(event));

    } catch (error) {
      console.error(error);
    }
  }
}

const eventAddNew = (event) => (
  {
    type: types.eventAddNew,
    payload: event
  }
);

export const eventStartUpdate = (event) => {
  return async (dispatch) => {
    try {
      const res = await fetchWithToken(`events/${event.id}`, event, 'PUT');
      const body = await res.json();

      if(body.ok) {
        dispatch(eventUpdated(event));
      } else {
        Swal.fire('Error', body.msg, 'error');
      }

    } catch (error) {
      console.error(error);
    }
  }
}

const eventUpdated = (event) => (
  {
    type: types.eventUpdated,
    payload: event
  }
);

export const eventStartDelete = () => {
  return async (dispatch, getState) => {
    try {
      const { id } = getState().calendar.activeEvent;
      const res = await fetchWithToken(`events/${id}`, {}, 'DELETE');
      const body = await res.json();
      if(body.ok) {
        dispatch(eventDeleted());
      } else {
        Swal.fire('Error', body.msg, 'error');
      }

    } catch (error) {
      console.error(error);
    }
  }
}

const eventDeleted = () => (
  {
    type: types.eventDeleted,
  }
);

export const eventClearActive = () => (
  {
    type: types.eventClearActive,
  }
);

export const eventStartLoad = () => {
  return async (dispatch) => {
    try {
      const res = await fetchWithToken('events');
      const body = await res.json();

      const events = convertDateEvents(body.events);
      dispatch(eventLoaded(events));

    } catch (error) {
      console.log(error);
    }
  }
}

const eventLoaded = (events) => (
  {
    type: types.eventStartLoad,
    payload: events
  }
)

export const eventClearState = () => (
  {
    type: types.eventClearState
  }
)