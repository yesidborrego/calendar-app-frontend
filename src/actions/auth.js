import Swal from "sweetalert2";

import { fetchWithoutToken, fetchWithToken } from "../helpers/fetch"
import { types } from "../types/types";
import { eventClearState } from "./events";

export const startLogin = (email, password) => {
  return async ( dispatch ) => {
    const res = await fetchWithoutToken('auth', {email, password}, 'POST');
    const body = await res.json();

    if( body.ok ) {
      localStorage.setItem('token', body.token);
      localStorage.setItem('token-init-date', new Date().getTime());

      dispatch(login({ uid: body.uid, name: body.name }));
    } else {
      Swal.fire('Error', body.msg, 'error');
    }
  }
}

export const startRegister = (email, password, name) => {
  return async (dispatch) => {
    const res = await fetchWithoutToken('auth/new', {email, password, name}, 'POST');
    const body = await res.json();

    if( body.ok ) {
      localStorage.setItem('token', body.token);
      localStorage.setItem('token-init-date', new Date().getTime());

      dispatch(login({ uid: body.uid, name: body.name }));
    } else {
      Swal.fire('Error', body.msg, 'error');
    }
  }
}

const login = (user) => (
  {
    type: types.authLogin,
    payload: user
  }
);

export const startTokenRenew = () => {
  return async (dispatch) => {
    const res = await fetchWithToken('auth/renew');
    const body = await res.json();

    if( body.ok ) {
      localStorage.setItem('token', body.token);
      localStorage.setItem('token-init-date', new Date().getTime());

      dispatch(login({ uid: body.uid, name: body.name }));
    } else {
      dispatch(authCheckingFinish());
    }
  }
}

const authCheckingFinish = () => ({
  type: types.authStartTokenRenew
})

export const startLogout = () => {
  return (dispatch) => {
    localStorage.clear();
    dispatch(eventClearState());
    dispatch(logout());
  }
}

const logout = () => (
  {
    type: types.authLogout
  }
);