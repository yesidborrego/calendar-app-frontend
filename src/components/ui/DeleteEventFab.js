import React from 'react'
import { useDispatch } from 'react-redux'

import { eventStartDelete } from '../../actions/events';

export const DeleteEventFab = () => {
  const dispatch = useDispatch();

  const handleDeleteEvent = () => {
    dispatch(eventStartDelete());
  }

  return (
    <button
      className="btn btn-danger fab-danger"
      onClick={ handleDeleteEvent }
    >
      <i className="fas fa-trash"></i>
      <span> Borrar evento </span>
    </button>
  )
}
