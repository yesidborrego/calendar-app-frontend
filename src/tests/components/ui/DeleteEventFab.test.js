import { mount } from 'enzyme';
import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import { eventStartDelete } from '../../../actions/events';
import { DeleteEventFab } from '../../../components/ui/DeleteEventFab';

import '@testing-library/jest-dom';

const middleware = [thunk];
const mockStore = configureStore(middleware);

const store = mockStore({});

store.dispatch = jest.fn();

jest.mock('../../../actions/events', () => ({
  eventStartDelete: jest.fn()
}));

const wrapper = mount(
  <Provider store={ store }>
    <DeleteEventFab/>
  </Provider>
);

describe('Probar el componente <DeleteEventFab />', () => {
  test('Debe mostrarse correctamente', () => {
    expect(wrapper).toMatchSnapshot();
  })

  test('Debe disparar el eventStartDelete', () => {
    wrapper.find('button').simulate('click');

    expect(eventStartDelete).toHaveBeenCalled();
  })

})
