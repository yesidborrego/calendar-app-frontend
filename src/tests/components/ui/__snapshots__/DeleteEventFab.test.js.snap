// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Probar el componente <DeleteEventFab /> Debe mostrarse correctamente 1`] = `
<button
  className="btn btn-danger fab-danger"
  onClick={[Function]}
>
  <i
    className="fas fa-trash"
  />
  <span>
     Borrar evento 
  </span>
</button>
`;
